src/ore/src/result.rs

```diff
@@ -37,6 +37,14 @@ pub trait ResultExt<T, E> {
     fn map_err_to_string(self) -> Result<T, String>
     where
         E: std::fmt::Display;
+
+    /// Safely unwraps a `Result<T, Infallible>`, where [`Infallible`] is a type that represents when
+    /// an error cannot occur.
+    ///
+    /// [`Infallible`]: core::convert::Infallible
+    fn infallible_unwrap(self) -> T
+    where
+        E: Into<core::convert::Infallible> + From<core::convert::Infallible>;
```

Comment:
I don't understand these bounds! It needs to be convertible both _into_ and _from_ `!`?

Reply:
Right! What I was trying to mimic here is an "Identity trait".

Ideally we only implement `infallible_unwrap` for `Result<T, Infallible>`, but `ResultExt` is implemented generically for any `Result<T, E>`. To avoid creating a second trait, e.g. `ResultInfallibleExt`, we need to add some bounds to `E` that only `Infallible` itself could satisfy, thus requiring it impl `Into` and `From`.

Requiring just the `Into` bound should be good enough though since it's impossible to create `Infallible`, as it's an [enum with no variants](https://doc.rust-lang.org/beta/src/core/convert/mod.rs.html#873). So the only way to implement `Into` would be something like:

```
impl Into<Infallible> for MyType {
  fn into(self) -> Infallible {
    todo!("can't create infallible")
  }
}
```

Updated!